Namespace-level configurable impersonation (programs)

anwar6953 opened the pull request:
Allow users to configure a principal and keytab (on local fs or hdfs), for which to launch cdap programs as.

https://issues.cask.co/browse/CDAP-6131
http://builds.cask.co/browse/CDAP-DUT4304
Integration tests: http://builds.cask.co/browse/CDAP-ITM-9

Note: In order to use this feature, https://github.com/caskdata/cdap/pull/5934 must be merged, or the permissions on the system hbase namespace must be granted to the impersonated principals.

anwar6953 committed:
Namespace-level configurable impersonation.

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/AbstractProgramRuntimeService.java:
```diff
@@ -118,7 +118,7 @@ public final RuntimeInfo run(ProgramDescriptor programDescriptor, ProgramOptions
     } catch (Exception e) {
       cleanUpTask.run();
-      LOG.error("Exception while trying to createPluginSnapshot", e);
+      LOG.error("Exception while trying to run program", e);
```
not your change, but why log and throw?

anwar6953 replied:
I'm not sure why it was done that way.
It's possible that the calling code doesn't log it the same way (since its from http handler).

bdmogal replied:
I'd say let's remove? If its from an HTTP Handler, then it will be captured by [HttpExceptionHandler](https://github.com/caskdata/cdap/blob/develop/cdap-common/src/main/java/co/cask/cdap/common/HttpExceptionHandler.java#L49)?

anwar6953 commented on cdap-proto/src/main/java/co/cask/cdap/proto/NamespaceMeta.java:
```diff
@@ -98,7 +114,15 @@ public NamespaceMeta build() {
       }
-      return new NamespaceMeta(name, description, new NamespaceConfig(schedulerQueueName));
+
+      if (principal == null) {
```
the pattern in NamespaceConfig is to default to empty strings

bdmogal replied:
should we change that? it seems strange

anwar6953 replied:
yea, I prefer nullable parameters also.

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractDistributedProgramRunner.java:
```diff
@@ -157,19 +168,46 @@ protected EventHandler createEventHandler(CConfiguration cConf) {
+  // returns the path of the file, if its a local URI. If it's an HDFS URI, copy it to local file system first.
+  private String getKeytabPath(URI keytabURI, File tempDir) throws IOException {
+    if (!"hdfs".equals(keytabURI.getScheme())) {
+      return keytabURI.getPath();
```
can't it be a non-hdfs, non-file path?

anwar6953 replied:
I think we can support it, if LocationFactory can support it. Specifically, if FileContextLocationFactory can support it.

bdmogal replied:
but then it may return a non-local path here?

bdmogal replied:
this still needs addressing, even though it became outdated.

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractDistributedProgramRunner.java:
```diff
@@ -196,100 +234,127 @@ public final ProgramController run(final Program program, final ProgramOptions o
+        String localizedKeytabPath = getKeytabPath(URI.create(keytabPath), tempDir);
+        LOG.info("Logging in as: principal={}, keytab={}", expandedPrincipal, localizedKeytabPath);
+
+        UserGroupInformation aliUGI =
```
change :)

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractDistributedProgramRunner.java:
```diff
@@ -196,100 +234,127 @@ public final ProgramController run(final Program program, final ProgramOptions o
+        return ProgramRunners.runAsUGI(aliUGI, callable);
+      } else {
+        // otherwise, simply launch as the current user
+        return callable.call();
```
nit: reverse the condition on line 341 to just have this in the if, so you reduce the indentation?

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ImpersonationUserResolver.java:
```diff
@@ -0,0 +1,79 @@
+  private final String defaultKeytabPath;
+
+  /**
+   * Construct SchedulerQueueResolver with CConfiguration and Store.
```
update javadoc (or remove it, doesn't seem like its necessary)

bdmogal commented on cdap-common/src/main/java/co/cask/cdap/common/kerberos/SecurityUtil.java:
```diff
@@ -142,12 +148,14 @@ public static boolean isKerberosEnabled(CConfiguration cConf) {
 
     if (UserGroupInformation.isSecurityEnabled()) {
+      validateKeytabPath(keytabPath);
+      String expandedPrincipal = SecurityUtil.expandPrincipal(principal);
```
don't need `SecurityUtil`?

bdmogal commented on cdap-common/src/main/java/org/apache/twill/internal/yarn/Hadoop21YarnAppClient.java:
```diff
@@ -50,21 +54,44 @@
+        yarnClients.put(userName, yarnClient);
+      }
+      return yarnClients.get(userName);
+    } catch (IOException e) {
```
seems like only line 76 should be caught?

anwar6953 replied:
true, but then I need to declare the variable outside the try block, so I can un-ident at that cost

bdmogal commented on cdap-common/src/main/java/org/apache/twill/internal/yarn/Hadoop21YarnAppClient.java:
```diff
@@ -50,21 +54,44 @@
+    this.configuration = configuration;
+  }
+
+  private YarnClient getYarnClient() {
```
startAndGetYarnClient?

bdmogal commented on cdap-common/src/main/java/org/apache/twill/internal/yarn/Hadoop21YarnAppClient.java:
```diff
@@ -126,22 +154,25 @@ private Resource adjustMemory(GetNewApplicationResponse response, Resource capab
   @Override
   protected void startUp() throws Exception {
-    yarnClient.start();
+    // no-op
```
perhaps add a note here on why you don't need to do anything here. otherwise, move that comment to getYarnClient and remove this method?

anwar6953 replied:
can't remove this method, since the class extends AbstractIdleService.
will update comment: `// no-op. We will create and start YarnClients, on demand`

bdmogal commented on cdap-proto/src/main/java/co/cask/cdap/proto/NamespaceConfig.java:
```diff
@@ -44,19 +58,33 @@ public boolean equals(Object o) {
+
+    NamespaceConfig that = (NamespaceConfig) o;
+
+    if (!schedulerQueueName.equals(that.schedulerQueueName)) {
```
just group line 64-70 in a single return statement?

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/NamespaceHttpHandler.java:
```diff
@@ -113,8 +113,17 @@ public void create(HttpRequest request, HttpResponder responder, @PathParam("nam
+        if (!Strings.isNullOrEmpty(config.getSchedulerQueueName())) {
+          builder.setSchedulerQueueName(config.getSchedulerQueueName());
+        }
+        // both or none of the following two must be set
```
do we enforce this constraint? if so, it seems like the if condition on line 124 should be nested inside the if condition on line 121?

anwar6953 replied:
its not enforced or required currently; removed the comment.
It was just a potential requirement we could have

On Tue, Jul 5, 2016 at 1:32 PM, Bhooshan Mogal notifications@github.com
wrote:

> In
> cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/NamespaceHttpHandler.java
> https://github.com/caskdata/cdap/pull/5975#discussion_r69632762:
> 
> > @@ -113,8 +113,17 @@ public void create(HttpRequest request, HttpResponder responder, @PathParam("nam
> >        }
> > 
> > ```
> >    NamespaceConfig config = metadata.getConfig();
> > ```
> > -      if (config != null && !Strings.isNullOrEmpty(config.getSchedulerQueueName())) {
> > -        builder.setSchedulerQueueName(config.getSchedulerQueueName());
> > -      if (config != null) {
> > -        if (!Strings.isNullOrEmpty(config.getSchedulerQueueName())) {
> > -          builder.setSchedulerQueueName(config.getSchedulerQueueName());
> > -        }
> > -        // both or none of the following two must be set
> 
> do we enforce this constraint? if so, it seems like the if condition on
> line 124 should be nested inside the if condition on line 121?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/caskdata/cdap/pull/5975/files/2b51590a6d3be9ad56fdd805040db06034e37800#r69632762,
> or mute the thread
> https://github.com/notifications/unsubscribe/ACU_ETXilcQ_9sgQFKLXi61HcGMXSwmSks5qSr9ygaJpZM4JFe4X
> .

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractDistributedProgramRunner.java:
```diff
@@ -157,19 +168,46 @@ protected EventHandler createEventHandler(CConfiguration cConf) {
+    // copy from HDFS to this local file
+    LocationFactory locationFactory = new FileContextLocationFactory(hConf);
+    final Location location = locationFactory.create(keytabURI);
+    LOG.info("Copying keytab file from {} to {}", location, localKeytabFile);
```
how often does this happen? once during program start?

anwar6953 replied:
yea

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractDistributedProgramRunner.java:
```diff
@@ -196,100 +234,127 @@ public final ProgramController run(final Program program, final ProgramOptions o
+
+          return launch(program, options, localizeResources, tempDir, new ApplicationLauncher() {
+            @Override
+            public TwillController launch(TwillApplication twillApplication, Iterable<String> extraClassPaths,
```
nothing has changed in this method, right?

anwar6953 replied:
yea, just indentation to wrap it in a callabel

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ImpersonationUserResolver.java:
```diff
@@ -0,0 +1,79 @@
+  /**
+   * Get Impersonated user at namespace level. If it is empty, returns the default principal.
+   *
+   * @param programId NamespaceId
```
programId, not namespaceId

anwar6953 replied:
variable name is self-descriptive, so removed the line

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ImpersonationUserResolver.java:
```diff
@@ -0,0 +1,79 @@
+/**
+ * Helper class to resolve the principal which CDAP will launch programs as.
+ */
+class ImpersonationUserResolver {
```
ImpersonatedUserResolver? Actually, should the class name care about impersonation? Just UserResolver?

anwar6953 replied:
ImpersonatedUserResolver sounded a bit stranger than the current name.
I don't like `UserResolver`, because its not ordinary user information.

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ImpersonationUserResolver.java:
```diff
@@ -0,0 +1,79 @@
+   * @param programId NamespaceId
+   * @return configured principal.
+   */
+  String getPrincipal(Id.Program programId) {
```
let's use the new id classes in this class.

anwar6953 committed:
Address comments

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractDistributedProgramRunner.java:
```diff
@@ -171,6 +171,10 @@ protected EventHandler createEventHandler(CConfiguration cConf) {
     if (!"hdfs".equals(keytabURI.getScheme())) {
+      // if it's not hdfs, assume its local file
+      if (keytabURI.getScheme() != null && !"file".equals(keytabURI.getScheme())) {
+        throw new IllegalArgumentException("Currently only file:// and hdfs:// URIs are supported.");
```
"... for keytab paths"? Would be good to mention the remedial action for this situation too.

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ImpersonationUserResolver.java:
```diff
@@ -46,11 +46,11 @@
    */
-  String getPrincipal(Id.Program programId) {
-    String namespaceUserSetting = getNamespaceConfig(programId.getNamespace()).getPrincipal();
+  @Nullable
```
how come this is nullable?

anwar6953 replied:
It could be the case that the cconf doesn't have the property defined.
Instead, I'll update the calling code to only call this if security is enabled.

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ImpersonationUserResolver.java:
```diff
@@ -60,20 +60,21 @@ String getPrincipal(Id.Program programId) {
    */
-  String getKeytabPath(Id.Program programId) {
-    String keytabPathSetting = getNamespaceConfig(programId.getNamespace()).getKeytabPath();
+  @Nullable
```
same question

anwar6953 replied:
see https://github.com/caskdata/cdap/pull/5975#discussion_r69978827

bdmogal commented on cdap-proto/src/main/java/co/cask/cdap/proto/NamespaceConfig.java:
```diff
@@ -61,13 +61,9 @@ public boolean equals(Object o) {
-      return false;
-    }
-    return keytabPath.equals(that.keytabPath);
+    return schedulerQueueName.equals(that.schedulerQueueName)
```
nit: we're using Objects.equals everywhere, right?

anwar6953 replied:
updated. Did not have Guava's equals/hashcode plugin, since I updated intellij to 16.

Edit: It can't use guava's equals/hashcode, since its in cdap-proto. Instead, I updated it to use java.util.Objects#equals and hashcode methods.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/PropertiesResolver.java:
```diff
@@ -35,11 +35,13 @@
   PropertiesResolver(PreferencesStore prefStore, NamespaceStore store, CConfiguration cConf) {
     this.prefStore = prefStore;
     this.queueResolver = new SchedulerQueueResolver(cConf, store);
+    this.impersonationUserResolver = new ImpersonationUserResolver(cConf, store);
```
Why new it? The `ImpersonationUserResolver` should have a `@Inject` constructor. BTW, will it be possible that there can be more one implementation of the `ImpersonationUserResolver`?

anwar6953 replied:
updated.
I can't think of another implementation of the class.

chtyim commented on cdap-common/src/main/java/org/apache/twill/internal/yarn/Hadoop21YarnAppClient.java:
```diff
@@ -50,21 +54,46 @@
  * </p>
  */
 // TODO: This is a copy from Twill, and can be removed when TWILL-119 is fixed. CDAP-4923 tracks Twill-0.8.0 upgrade.
+// After copying, we have removed the addRMToken() method. That functionality is in YarnTokenUtils.
```
Since this is a class copied from Twill, do you have a corresponding twill ticket? You should include it in the TODO as well.

anwar6953 replied:
updated.

chtyim commented on cdap-proto/src/main/java/co/cask/cdap/proto/NamespaceConfig.java:
```diff
@@ -28,14 +26,30 @@
   private final String schedulerQueueName;
 
-  public NamespaceConfig(String schedulerQueueName) {
+  @SerializedName("principal")
```
Do we always apply the `@serializedName`? Normally we don't use it and only use it for backward compatible reason.

anwar6953 replied:
removed from the new fields.

Edit: actually, its used so that the user-facing key is something like 'keytab.path', instead of keytabPath.

bdmogal replied:
but like @chtyim said, its not usually used, let's not add it to newer fields?

anwar6953 replied:
from the user perspective, `keytab.path` may be more like a configuration parameter than `keytabPath`.

bdmogal replied:
but we only do this for NamespaceConfig, right. No other proto classes have this style, so lets standardize?

anwar6953 replied:
done

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractDistributedProgramRunner.java:
```diff
@@ -157,19 +168,50 @@ protected EventHandler createEventHandler(CConfiguration cConf) {
     return new AbortOnTimeoutEventHandler(cConf.getLong(Constants.CFG_TWILL_NO_CONTAINER_TIMEOUT, Long.MAX_VALUE));
   }
 
+  // returns the path of the file, if its a local URI. If it's an HDFS URI, copy it to local file system first.
```
Better use the `Locations.linkOrCopy` method.

anwar6953 replied:
can't use that, because it doesn't allow make the new file to have restricted permissions.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractDistributedProgramRunner.java:
```diff
@@ -157,19 +168,50 @@ protected EventHandler createEventHandler(CConfiguration cConf) {
+  // returns the path of the file, if its a local URI. If it's an HDFS URI, copy it to local file system first.
+  private String getKeytabPath(URI keytabURI, File tempDir) throws IOException {
+    if (!"hdfs".equals(keytabURI.getScheme())) {
+      // if it's not hdfs, assume its local file
```
This is not true (`maprfs`, MS azure FS).

anwar6953 replied:
Will `FileContextLocationFactory` automatically handle those?

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractDistributedProgramRunner.java:
```diff
@@ -157,19 +168,50 @@ protected EventHandler createEventHandler(CConfiguration cConf) {
+    final Location location = locationFactory.create(keytabURI);
+    LOG.info("Copying keytab file from {} to {}", location, localKeytabFile);
+    try (InputStream is = location.getInputStream()) {
+      ByteStreams.copy(is, Files.newOutputStreamSupplier(localKeytabFile));
```
Use the new Java `Files.copy` API instead.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractDistributedProgramRunner.java:
```diff
@@ -157,19 +168,50 @@ protected EventHandler createEventHandler(CConfiguration cConf) {
+    }
+
+    // create a local file with restricted permissions
+    File localKeytabFile = new File(tempDir, "keytab.localized");
```
Use the new Java `Files.createFile` instead so that you can set the right permission when the file is created.

chtyim commented on cdap-common/src/main/java/org/apache/twill/internal/yarn/Hadoop21YarnAppClient.java:
```diff
@@ -50,21 +54,46 @@
+
+  // TODO: make thread safe
+  // alternatively, simply construct a new YARNClient when needed. Shouldn't be too expensive
+  private Map<String, YarnClient> yarnClients = new HashMap<>();
```
Better not to cache.

anwar6953 replied:
It's fine not to call YarnClient#stop(), then?

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/PropertiesResolver.java:
```diff
@@ -36,12 +37,17 @@
-    this.impersonationUserResolver = new ImpersonationUserResolver(cConf, store);
+    this.queueResolver = schedulerQueueResolver;
+    this.impersonationUserResolver = impersonationUserResolver;
+
```
nit: remove empty newline

anwar6953 committed:
Fix test cases

anwar6953 commented:
@bdmogal
Pushed one more commit to fix test cases.
Will merge after build succeeds.

anwar6953 commented:
Addressed comments, but closing this PR, as a follow-up PR has significant refactoring, of this code. So just keeping that entire PR for review: https://github.com/caskdata/cdap/pull/6024
